File: music_manager.go
package main

import (
	"encoding/binary"
	"io"
	"io/ioutil"
	"log"
	"math"
	"math/rand"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// ================= 🎵 音乐管理器 =================
type MusicManager struct {
	isPlaying     bool
	mu            sync.Mutex
	cmd           *exec.Cmd
	stdin         io.WriteCloser
	stopChan      chan struct{}
	targetVolume  float64
	currentVolume float64
	volMutex      sync.Mutex
}

func NewMusicManager() *MusicManager {
	return &MusicManager{targetVolume: 1.0, currentVolume: 1.0}
}

func (m *MusicManager) IsPlaying() bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.isPlaying
}

func (m *MusicManager) setTargetVolume(vol float64) {
	m.volMutex.Lock()
	m.targetVolume = vol
	m.volMutex.Unlock()
}

func (m *MusicManager) Duck() {
	if m.IsPlaying() {
		// Duck 要“明显且快速”：先把目标压到 20%，并把当前音量立即拉到一个较低上限，
		// 避免因为缓慢平滑导致用户听感“没有降音量”。
		m.volMutex.Lock()
		m.targetVolume = 0.2
		if m.currentVolume > 0.35 {
			m.currentVolume = 0.35
		}
		m.volMutex.Unlock()
	}
}

func (m *MusicManager) Unduck() {
	if m.IsPlaying() {
		m.setTargetVolume(1.0)
	}
}

func (m *MusicManager) Stop() {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.isPlaying {
		log.Println("🛑 [MUSIC] 停止播放")
		select {
		case m.stopChan <- struct{}{}:
		default:
		}
		if m.stdin != nil {
			m.stdin.Close()
		}
		if m.cmd != nil && m.cmd.Process != nil {
			m.cmd.Process.Kill()
			m.cmd.Wait()
		}
		m.isPlaying = false
	}
}

func (m *MusicManager) PlayFile(path string) {
	m.Stop()
	time.Sleep(200 * time.Millisecond)

	m.mu.Lock()
	defer m.mu.Unlock()

	file, err := os.Open(path)
	if err != nil {
		return
	}

	// -B 是缓冲时间(us)：太小会在 CPU 抖动时 underrun（卡顿），太大会导致 Duck/切歌响应滞后。
	// 这里取一个折中值，配合下游“前置缓冲”控制，保证不卡顿且仍可及时 Duck。
	cmd := exec.Command("aplay", "-D", "default", "-q", "-t", "raw", "-r", "16000", "-c", "1", "-f", "S16_LE", "-B", "80000")
	stdin, err := cmd.StdinPipe()
	if err != nil {
		file.Close()
		return
	}
	if err := cmd.Start(); err != nil {
		file.Close()
		return
	}

	m.cmd = cmd
	m.stdin = stdin
	m.isPlaying = true
	m.stopChan = make(chan struct{}, 1)
	m.targetVolume = 1.0
	m.currentVolume = 1.0

	log.Printf("🎵 [MUSIC] 开始播放: %s", filepath.Base(path))

	go func(f *os.File, pipe io.WriteCloser, myCmd *exec.Cmd, stopCh chan struct{}) {
		defer f.Close()
		defer pipe.Close()
		f.Seek(44, 0)
		// 关键：
		// - 不能“严格实时”地喂数据（每 20ms sleep 一次），否则在 RK3308 上只要调度抖动就会 underrun（听感卡顿）。
		// - 也不能一次性喂太快/太多，否则 Duck 的听感会滞后（因为旧音量的音频已经预灌进 aplay/管道）。
		//
		// 策略：维护一个小的“前置缓冲”（例如 120~180ms），既抗抖动又保证 Duck 仍然足够跟手。
		const (
			musicSampleRate = 16000
			chunkSamples    = 640 // 40ms：降低调度开销，同时仍有较好音量跟随
			targetAhead     = 120 * time.Millisecond
			maxAhead        = 180 * time.Millisecond
		)
		buf := make([]byte, chunkSamples*2)

		var (
			startWall    time.Time
			wroteSamples int64
			lastStepAt   time.Time
		)
		for {
			select {
			case <-stopCh:
				return
			default:
			}
			n, err := f.Read(buf)
			if n > 0 {
				if startWall.IsZero() {
					startWall = time.Now()
					lastStepAt = startWall
				}

				// 读取当前目标音量，并按 dt 做平滑逼近（Duck 快、Unduck 慢）
				now := time.Now()
				dt := now.Sub(lastStepAt)
				lastStepAt = now

				m.volMutex.Lock()
				target := m.targetVolume
				current := m.currentVolume
				m.volMutex.Unlock()

				if target < 0 {
					target = 0
				} else if target > 1 {
					target = 1
				}
				if current < 0 {
					current = 0
				} else if current > 1 {
					current = 1
				}

				if dt <= 0 {
					current = target
				} else if current != target {
					var tau time.Duration
					if target < current {
						tau = 120 * time.Millisecond
					} else {
						tau = 900 * time.Millisecond
					}
					alpha := 1 - math.Exp(-float64(dt)/float64(tau))
					if alpha < 0 {
						alpha = 0
					} else if alpha > 1 {
						alpha = 1
					}
					current = current + (target-current)*alpha
				}

				m.volMutex.Lock()
				m.currentVolume = current
				m.volMutex.Unlock()

				// PCM16 振幅缩放 + 饱和裁剪
				for i := 0; i+1 < n; i += 2 {
					sample := int16(binary.LittleEndian.Uint16(buf[i : i+2]))
					v := int(float64(sample) * current)
					if v > 32767 {
						v = 32767
					} else if v < -32768 {
						v = -32768
					}
					binary.LittleEndian.PutUint16(buf[i:i+2], uint16(int16(v)))
				}

				if _, werr := pipe.Write(buf[:n]); werr != nil {
					return
				}

				// 维护“前置缓冲”：若已写入的音频时长领先于墙钟太多，则主动 sleep 让播放追上来。
				wroteSamples += int64(n / 2)
				audioDur := time.Duration(wroteSamples) * time.Second / musicSampleRate
				ahead := audioDur - time.Since(startWall)
				if ahead > maxAhead {
					sleepDur := ahead - targetAhead
					if sleepDur > 0 {
						select {
						case <-stopCh:
							return
						case <-time.After(sleepDur):
						}
					}
				}
			}

			if err != nil {
				break
			}
		}
		m.mu.Lock()
		if m.isPlaying && m.cmd == myCmd {
			m.isPlaying = false
			go myCmd.Wait()
		}
		m.mu.Unlock()
	}(file, stdin, cmd, m.stopChan)
}

func (m *MusicManager) SearchAndPlay(query string) bool {
	files, err := ioutil.ReadDir(musicDir)
	if err != nil {
		return false
	}
	var candidates []string
	for _, f := range files {
		if !f.IsDir() && strings.HasSuffix(strings.ToLower(f.Name()), ".wav") {
			candidates = append(candidates, filepath.Join(musicDir, f.Name()))
		}
	}
	if len(candidates) == 0 {
		return false
	}
	target := ""
	if query == "RANDOM" {
		target = candidates[rand.Intn(len(candidates))]
	} else {
		q := strings.ToLower(query)
		for _, path := range candidates {
			if strings.Contains(strings.ToLower(filepath.Base(path)), q) {
				target = path
				break
			}
		}
		if target == "" {
			return false
		}
	}
	m.PlayFile(target)
	return true
}

File: py_ai_box/music_manager.py
import math
import os
import random
import subprocess
import threading
import time
from array import array
from pathlib import Path

from . import state
from . import config_runtime


class MusicManager:
    def __init__(self) -> None:
        self._is_playing = False
        self._mu = threading.Lock()
        self._cmd = None
        self._stdin = None
        self._stop_event = threading.Event()
        self._target_volume = 1.0
        self._current_volume = 1.0
        self._vol_lock = threading.Lock()

    def is_playing(self) -> bool:
        with self._mu:
            return self._is_playing

    def _set_target_volume(self, vol: float) -> None:
        with self._vol_lock:
            self._target_volume = vol

    def duck(self) -> None:
        if self.is_playing():
            with self._vol_lock:
                self._target_volume = 0.2
                if self._current_volume > 0.35:
                    self._current_volume = 0.35

    def unduck(self) -> None:
        if self.is_playing():
            self._set_target_volume(1.0)

    def stop(self) -> None:
        with self._mu:
            if self._is_playing:
                print("🛑 [MUSIC] 停止播放")
                self._stop_event.set()
                if self._stdin is not None:
                    try:
                        self._stdin.close()
                    except Exception:
                        pass
                if self._cmd is not None and self._cmd.poll() is None:
                    try:
                        self._cmd.kill()
                        self._cmd.wait()
                    except Exception:
                        pass
                self._is_playing = False
                self._stop_event = threading.Event()

    def play_file(self, path: str) -> None:
        self.stop()
        time.sleep(0.2)

        with self._mu:
            try:
                f = open(path, "rb")
            except Exception:
                print(f"[MUSIC-DBG] open_failed path={path}")
                return

            cmd = subprocess.Popen(
                ["aplay", "-D", "default", "-q", "-t", "raw", "-r", "16000", "-c", "1", "-f", "S16_LE", "-B", "80000"],
                stdin=subprocess.PIPE,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
            if cmd.stdin is None:
                f.close()
                print(f"[MUSIC-DBG] aplay_start_failed path={path}")
                return

            self._cmd = cmd
            self._stdin = cmd.stdin
            self._is_playing = True
            self._target_volume = 1.0
            self._current_volume = 1.0
            print(f"🎵 [MUSIC] 开始播放: {Path(path).name}")

            thread = threading.Thread(
                target=self._stream_file,
                args=(f, cmd, self._stop_event),
                daemon=True,
            )
            thread.start()

    def _stream_file(self, f, cmd, stop_event: threading.Event) -> None:
        try:
            f.seek(44)
            music_sample_rate = 16000
            chunk_samples = 640  # 40ms
            target_ahead = 0.12
            max_ahead = 0.18
            buf_bytes = chunk_samples * 2

            start_wall = None
            wrote_samples = 0
            last_step_at = None

            while not stop_event.is_set():
                data = f.read(buf_bytes)
                if not data:
                    break

                if start_wall is None:
                    start_wall = time.time()
                    last_step_at = start_wall

                now = time.time()
                dt = max(0.0, now - (last_step_at or now))
                last_step_at = now

                with self._vol_lock:
                    target = min(1.0, max(0.0, self._target_volume))
                    current = min(1.0, max(0.0, self._current_volume))

                if dt == 0:
                    current = target
                elif current != target:
                    tau = 0.12 if target < current else 0.9
                    alpha = 1 - math.exp(-dt / tau)
                    current = current + (target - current) * alpha

                with self._vol_lock:
                    self._current_volume = current

                samples = array("h")
                samples.frombytes(data)
                for i in range(len(samples)):
                    v = int(samples[i] * current)
                    if v > 32767:
                        v = 32767
                    elif v < -32768:
                        v = -32768
                    samples[i] = v

                try:
                    self._stdin.write(samples.tobytes())
                    self._stdin.flush()
                except Exception:
                    break

                wrote_samples += len(samples)
                audio_dur = wrote_samples / music_sample_rate
                ahead = audio_dur - (time.time() - start_wall)
                if ahead > max_ahead:
                    sleep_dur = ahead - target_ahead
                    if sleep_dur > 0:
                        time.sleep(sleep_dur)
        finally:
            f.close()
            with self._mu:
                if self._is_playing and self._cmd == cmd:
                    self._is_playing = False
                    try:
                        cmd.wait(timeout=1)
                    except Exception:
                        pass

    def search_and_play(self, query: str) -> bool:
        try:
            files = os.listdir(config_runtime.music_dir)
        except Exception as err:
            print(f"[MUSIC-DBG] list_dir_failed dir={config_runtime.music_dir} err={err}")
            return False
        candidates = []
        for name in files:
            if name.lower().endswith(".wav"):
                candidates.append(str(Path(config_runtime.music_dir) / name))
        if not candidates:
            print(f"[MUSIC-DBG] no_wav dir={config_runtime.music_dir}")
            return False
        target = ""
        if query == "RANDOM":
            target = random.choice(candidates)
        else:
            q = query.lower()
            for path in candidates:
                if q in Path(path).name.lower():
                    target = path
                    break
            if not target:
                print(f"[MUSIC-DBG] no_match query={query}")
                return False
        print(f"[MUSIC-DBG] match file={Path(target).name}")
        self.play_file(target)
        return True


def init_music_manager() -> MusicManager:
    mgr = MusicManager()
    state.music_mgr = mgr
    return mgr
